internal/mux/mux.go

```diff
@@ -78,33 +84,37 @@ func (m *Mux) Read(buf []byte) (int, error) {
 		}
 	}
 
-	bs := m.channelNum * m.bitDepthInBytes
-	l := len(buf)
-	l = l / bs * bs // Adjust the length in order not to mix different channels.
-
-	bufs := map[*bufio.Reader][]byte{}
-	for _, p := range m.readers {
-		peeked, err := p.Peek(l)
-		if err != nil && err != bufio.ErrBufferFull && err != io.EOF {
-			return 0, err
-		}
-		if l > len(peeked) {
-			l = len(peeked)
-			l = l / bs * bs
+	minBufferSize := len(buf)
+	if minBufferSize > 256 {
+		minBufferSize = 256
+	}
+	for buf := range m.buffers {
+		l := buf.Len()
+		if l > 0 && l < minBufferSize {
+			minBufferSize = l
 		}
-		bufs[p] = peeked[:l]
 	}
 
+	bs := m.channelNum * m.bitDepthInBytes
+	l := minBufferSize / bs * bs
+
 	if l == 0 {
 		// Returning 0 without error can block the caller of Read forever. Call Gosched to encourage context switching.
 		runtime.Gosched()
 		return 0, nil
 	}
 
-	for _, p := range m.readers {
-		if _, err := p.Discard(l); err != nil && err != io.EOF {
+	bufs := make([][]byte, 0, len(m.buffers))
+	for buf := range m.buffers {
+		if buf.Len() == 0 {
+			continue
+		}
+		sl := make([]byte, l)
+		_, err := buf.Read(sl)
+		if err != nil && err != io.EOF {
 			return 0, err
 		}
+		bufs = append(bufs, sl)
```

Comment:
Assuming that the read byte number is n, we must specify `sl[:n]`

Reply:
With the above swap to readfull, I think this would be unnecessary.